sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/CustomMetric.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.read;
+
+import org.apache.spark.annotation.Evolving;
+
+/**
+ * A general custom metric.
+ *
+ * @since 3.2.0
+ */
+@Evolving
+public interface CustomMetric {
```

Comment:
The comment https://github.com/apache/spark/pull/31451#discussion_r569774028 suggested to name it `LongMetric`. But later I think if we may need keep flexibility in the base interface.

So I left it as general as possible and add a `LongMetric` which reports a long value.